R/guides-.R

```diff
@@ -485,15 +485,76 @@ Guides <- ggproto(
       return(zeroGrob())
     }
 
+    # extract the guide position
+    positions <- vapply(
+      params,
+      function(p) p$position[1] %||% default_position,
+      character(1), USE.NAMES = FALSE
+    )
+
     # Populate key sizes
     theme$legend.key.width  <- calc_element("legend.key.width",  theme)
     theme$legend.key.height <- calc_element("legend.key.height", theme)
 
-    grobs <- self$draw(theme, default_position, theme$legend.direction)
+    grobs <- self$draw(theme, positions, theme$legend.direction)
+    keep <- !vapply(grobs, is.zero, logical(1), USE.NAMES = FALSE)
+    grobs <- grobs[keep]
     if (length(grobs) < 1) {
       return(zeroGrob())
     }
-    grobs <- grobs[order(names(grobs))]
+
+    # prepare the position of inside legends
+    default_inside_just <- valid.just(
+      calc_element("legend.justification.inside", theme)
+    )
+    default_inside_position <- calc_element(
+      "legend.position.inside", theme
+    )
+    inside_justs <- inside_positions <- vector("list", length(positions))
+
+    # we grouped the legends by the positions, for inside legends, they'll be
+    # splitted by the actual inside coordinate
+    groups <- positions
+    for (i in seq_along(positions)) {
+      if (identical(positions[i], "inside")) {
+        # the actual inside position and justification can be set in each guide
+        # by `theme` argument, here, we won't use `calc_element()` which will
+        # use inherits from `legend.justification` or `legend.position`, we only
+        # follow the inside elements from the guide theme
+        inside_just <- params[[i]]$theme[["legend.justification.inside"]]
+        if (is.null(inside_just)) {
+          inside_justs[[i]] <- default_inside_just
+        } else {
+          inside_justs[[i]] <- valid.just(inside_just)
+        }
+        inside_positions[[i]] <- params[[i]]$theme[[
+          "legend.position.inside"
+        ]] %||% default_inside_position %||% inside_justs[[i]]
+        groups[i] <- paste("inside", 
+          paste(inside_positions[[i]], collapse = "_"),
+          paste(inside_justs[[i]], collapse = "_"),
+          sep = "_"
+        )
+      }
+    }
+
+    positions <- positions[keep]
+    inside_positions <- inside_positions[keep]
+    inside_justs <- inside_justs[keep]
+    groups <- groups[keep]
+
+    # we group the guide legends
+    locs <- vec_group_loc(groups)
+    indices <- locs$loc
+    grobs <- vec_chop(grobs, indices = indices)
+    names(grobs) <- locs$key
+
+    # for each group, they share the same locations,
+    # so we only extract the first one of `positions` and `inside_positions`
+    first_indice <- lapply(indices, `[[`, 1L)
+    positions <- vec_chop(positions, indices = first_indice)
+    inside_positions <- vec_chop(inside_positions, indices = first_indice)
+    inside_justs <- vec_chop(inside_justs, indices = first_indice)
```

Comment:
I think this could be simplified by tracking positions/inside_positions/inside_justs in a data.frame.
`vec_group_loc()` treats rows that are the same as the same group, and the `locs$key` should allow you to skip having the subset the first index.

Reply:
Thanks for your suggestions! Should be fixed